### DISCUSSION TITLE
iOS permissions not working in Appium & Robot Framework test automation

### DISCUSSION THREAD

**Discussion by espi66** [appium/appium#15337] [Q&A]
I'm trying automate a native application which is running in AWS's Device Farm. Appium server is 1.19.0 and I've tried with later versions too (up to 1.21.0), but when accepting location in Settings it does not change location permission for the app. This applies only to iOS 13 series. When I try it with iOS 14 series, the picture gallery permission is not permitted for the app. Only iOS versions prior to 13 work beautifully.
So, is there a way to handle permissions in iOS devices so that they can be permitted for the app while doing test automation with Robot Framework and Appium, also in later iOS versions? Or is Appium server not supporting latest iOS software well enough yet? Or has Apple limited test automation possibilities on purpose? Anyway, permissions work in manual testing so why shouldn't it work in test automation as well?

**Comment by jlipps** [comment]
Which capability or command are you using in the working scenario with iOS 13? I couldn't tell from what you wrote.

**Reply by espi66** [reply]
Application uses iOS's own Settings. Even if I'm navigating to Settings (app has a button linking there), correcting Location to 'When/While Using The App', and coming back to the app, the location setting is still not permitted to use. The same applies to picture gallery, and it's not even obligatory for the app to have permissions there.

**Reply by KazuCocoa** [reply]
It seems something iOS restricts behavior via XCTest. (It's my guess though)


As Appium/WDA, it does not inject iOS behavior since we cannot. Most behavior is via XCTest framework by Apple. If apple restricted some behavior via XCTest framework (it is a blackbox, so we only can observe the behavior and guess it)

Do you have the full appium log with `showIOSLog` capability? It may help to see some kind of log messages.

**Reply by espi66** [reply]
As I'm running it in AWS Device Farm, I think the access to meaningful logs is prevented from us users? If not, please advice.

**Reply by KazuCocoa** [reply]
I haven't used AWS device farm for iOS, so I have no good idea. If you can allow the permission via remote control (I'm not sure AWS provides this feature), potentially you can allow the permission via it. And can run Appium session against the device.

The behavior seems to a technical limitation by iOS over XCTest connection. Afaik, Siri also resets its preference via automation, not only Appium but also vanilla XCTest, in some cases. Then, changing preference outside the automation session was needed.

**Reply by espi66** [reply]
I think I managed to get the logs out. It's a huge file, of several megas, maybe not possible to send it here?

**Reply by KazuCocoa** [reply]
You probably can use https://gist.github.com/

**Comment by espi66** [comment]
It takes too long to load the page, the file is far too large.
Could I send it via this email address?

**Reply by KazuCocoa** [reply]
yes. You can find my email address in my GitHub profile
